File: clients/ruby/model/vec2.rb
module Model

# 2 dimensional vector.
class Vec2
    # `x` coordinate of the vector
    attr_accessor :x
    # `y` coordinate of the vector
    attr_accessor :y

    def initialize(x, y)
        @x = x
        @y = y
    end

    # Read Vec2 from input stream
    def self.read_from(stream)
        x = stream.read_double()
        y = stream.read_double()
        Vec2.new(x, y)
    end

    # Write Vec2 to output stream
    def write_to(stream)
        stream.write_double(@x)
        stream.write_double(@y)
    end

    def to_s
        string_result = "Vec2 { "
        string_result += "x: "
        string_result += @x.to_s
        string_result += ", "
        string_result += "y: "
        string_result += @y.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end

File: clients/go/model/vec2.go
package model

import (
	"fmt"
	"io"

	"aicup22/stream"
)

// 2 dimensional vector.
type Vec2 struct {
	// `x` coordinate of the vector
	X float64
	// `y` coordinate of the vector
	Y float64
}

func NewVec2(x, y float64) Vec2 {
	return Vec2{
		X: x,
		Y: y,
	}
}

// Read Vec2 from reader
func ReadVec2(reader io.Reader) Vec2 {
	x := stream.ReadFloat64(reader)
	y := stream.ReadFloat64(reader)
	return Vec2{
		X: x,
		Y: y,
	}
}

// Write Vec2 to writer
func (vec2 Vec2) Write(writer io.Writer) {
	x := vec2.X
	stream.WriteFloat64(writer, x)
	y := vec2.Y
	stream.WriteFloat64(writer, y)
}

// Get string representation of Vec2
func (vec2 Vec2) String() string {
	stringResult := "{ "
	stringResult += "X: "
	x := vec2.X
	stringResult += fmt.Sprint(x)
	stringResult += ", "
	stringResult += "Y: "
	y := vec2.Y
	stringResult += fmt.Sprint(y)
	stringResult += " }"
	return stringResult
}
